contracts/token/ERC20/ERC20Snapshot.sol

```diff
@@ -0,0 +1,130 @@
+pragma solidity ^0.4.24;
+
+import "./ERC20.sol";
+import "../../utils/Arrays.sol";
+
+
+/**
+ * @title ERC20Snapshot token
+ * @dev An ERC20 token which enables taking snapshots of account balances.
+ * This can be useful to safely implement voting weighed by balance.
+ */
+contract ERC20Snapshot is ERC20 {
+  
+  using Arrays for uint256[];
+
+  // The 0 id represents no snapshot was taken yet.
+  uint256 private currentSnapshotId;
+
+  mapping (address => uint256[]) private snapshotIds;
+  mapping (address => uint256[]) private snapshotBalances;
+
+  event Snapshot(uint256 id);
+
+  /**
+   * @dev Increments current snapshot. Emites Snapshot event.
+   * @return An uint256 representing current snapshot id.
+   */
+  function snapshot() external returns (uint256) {
+    currentSnapshotId += 1;
+    emit Snapshot(currentSnapshotId);
+    return currentSnapshotId;
+  }
+
+  /**
+   * @dev Returns account balance for specific snapshot.
+   * @param account address The address to query the balance of.
+   * @param snapshotId uint256 The snapshot id for which to query the balance of.
+   * @return An uint256 representing the amount owned by the passed address for specific snapshot.
+   */
+  function balanceOfAt(
+    address account,
+    uint256 snapshotId
+  )
+    public
+    view
+    returns (uint256)
+  {
+    require(snapshotId > 0 && snapshotId <= currentSnapshotId);
+
+    uint256 idx = snapshotIds[account].findUpperBound(snapshotId);
+
+    if (idx == snapshotIds[account].length) {
+      return balanceOf(account);
+    } else {
+      return snapshotBalances[account][idx];
+    }
+  }
+
+  /**
+  * @dev Transfer token for a specified address. It takes balance snapshot for the sender and recipient account
+  * before transfer is done.
+  * @param to The address to transfer to.
+  * @param value The amount to be transferred.
+  */
+  function transfer(address to, uint256 value) public returns (bool) {
+    updateSnapshot(msg.sender);
+    updateSnapshot(to);
+    return super.transfer(to, value);
```

Comment:
`super.transfer()` should be called first to follow OZ's standard of failing early. 
Also the `balanceOf()` function call made below after calling `updateSnapshot()` will not be correct as the state change has not occurred yet.